### PR TITLE
[X86] SimplifyDemandedVectorEltsForTargetNode - reduce width of X86 conversions nodes when upper elements are not demanded.

### LINKED ISSUE #83402: [X86] v4f16 -> v4i32 conversion unnecessarily using YMM registers
https://rust.godbolt.org/z/PEfxzYY4z

```ll
define <4 x i32> @fptosi_4f16_to_4i32(<4 x half> %a) nounwind {
  %cvt = fptosi <4 x half> %a to <4 x i32>
  ret <4 x i32> %cvt
}
```
llc -mcpu=x86-64-v3
```asm
fptosi_4f16_to_4i32:                    # @fptosi_4f16_to_4i32
        vcvtph2ps       %xmm0, %ymm0
        vcvttps2dq      %ymm0, %ymm0
        vzeroupper
        retq
```

We should only require the xmm variants (and avoid the vzeroupper entirely)

### PR DESCRIPTION
**Pull request by RKSimon**
Fixes #83402

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -42518,6 +42518,25 @@ bool X86TargetLowering::SimplifyDemandedVectorEltsForTargetNode(
       SDValue Insert =
           insertSubVector(UndefVec, ExtOp, 0, TLO.DAG, DL, ExtSizeInBits);
       return TLO.CombineTo(Op, Insert);
+    }
+      // Conversions.
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:42522]: Do we need to consider other conversion nodes like `CVTP2SI/CVTP2UI/CVTSI2P/VFPROUND...`?

Comment (RKSimon) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:42522]: Absolutely, I was thinking of adding them when I have test cases - but I can try to add them all now if you prefer?

Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:42522]: I'm OK with adding it later with test case. Maybe leave a comment here for now.

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**